doctor/lib/ocr_utils.py

```diff
@@ -0,0 +1,345 @@
+    """Average character width for a block of text
+
+    :param block_data:
+    :return: average character width
+    """
+    fd = block_data[block_data.text.str.len() > 0]
+    return (fd.width / fd.text.str.len()).mean()
+
+
+def validate_ocr_text(row: pd.Series, img: Image) -> pd.Series:
+    """Validate the OCR results
+
+    Take a look at our OCR and remove the bad results
+    and if possible reprocess words one by one
+
+    :param row: row of ocr results
+    :param img: Page image
+    :return: Updated row if necessary
+    """
+    if row["left"] < 370 and row["conf"] <= 40:
+        row["text"] = " " * len(row["text"])
+    elif row["conf"] < 10 and len(row["text"]) >= 3:
+        bbox = (
+            row["left"] - 5,
+            row["top"] - 3,
+            row["left"] + row["width"] + 5,
+            row["top"] + row["height"] + 3,
+        )
+        word_df = pd.DataFrame(
+            pytesseract.image_to_data(
+                img.crop(bbox),
+                config="-c preserve_interword_spaces=1x1 --psm 7 -l eng",
+                output_type=Output.DICT,
+            )
+        )
+        new_words = " ".join(
+            word_df.loc[word_df["conf"] > 10, "text"].tolist()
+        )
+        if new_words:
+            row["text"] = new_words
+        else:
+            row["text"] = "□" * len(row["text"])
+    elif row["conf"] < 10:
+        # if the confidence is under 10 and its just three characters - box it
+        row["text"] = "□"
+    elif (
+        row["conf"] < 20
+        and len(row["text"]) == 1
+        and not row["text"].isalnum()
+    ):
+        # Artifacts from scans often appear as lone symbols
+        # if conf is low and they are all alone drop them
+        row["text"] = " "
+
+    return row["text"] + " "
+
+
+def add_newlines(row: pd.Series, state: dict) -> dict:
+    """Add new linebreaks into the ocr'd page
+
+    Identify where line breaks should be added
+
+    :param row: the row of data from tesseract
+    :param state: the location data used to decide where line breaks should be
+    :return:
+    """
+    prev = state["prev_row"]
+    max_y = state["max_y"]
+    new_line = prev["line_num"] != row["line_num"] if max_y > 0 else True
+    new_paragraph = prev["par_num"] != row["par_num"] if max_y > 0 else True
+
+    if new_line:
+        state["page_text"] += "\n"
+        state["indent"] = 0
+    if new_paragraph:
+        # Add a second line break for new paragraphs for good measure
+        state["page_text"] += "\n"
+        state["indent"] = 0
+
+    if new_line and not new_paragraph and state["max_y"] > 0:
+        diff = row["top"] - state["max_y"]
+        if 200 > diff > 130:
+            state["page_text"] += "\n"
+        elif diff > 200:
+            state["page_text"] += "\n\n"
+        state["max_y"] = 0
+
+    state["max_y"] = max(state["max_y"], row["top"] + row["height"])
+    return state
+
+
+def insert_indentation(row: pd.Series, state: dict) -> dict:
```

Comment:
This is pretty tough to make sense of too. Maybe we need a note somewhere explaining what the tesseract dicts mean as well as some line by line comments explaining things?